### PR TITLE
No quadratic Get population

### PR DESCRIPTION
**Pull request by yihozhang**

### CODE REVIEW

File: dag_in_context/src/utility/util.egg [hunk 1]
```diff
@@ -25,35 +25,67 @@
        (= len (TypeList-length tl)))
       ((set (tuple-length expr) len)) :ruleset always-run)
 
+
+;; Leading expressions are tuples that are used as a whole
+;; during optimization and are matched in the query.
+(relation leading-Expr (Expr))
+(relation leading-Expr-list (ListExpr))
```
Comment (oflatt) [inline dag_in_context/src/utility/util.egg:32]: Is leading-expr-list even used anywhere? perhaps we should cut it?

File: tests/snapshots/files__implicit-return-optimize.snap [hunk 2]
```diff
@@ -53,31 +54,32 @@ expression: visualization.result
   v6_: int = id c1_;
   v7_: int = id c3_;
 .b8_:
-  c9_: int = const 14;
```
Comment (oflatt) [inline tests/snapshots/files__implicit-return-optimize.snap:56]: huh, I wonder why the old one has const 14 while the new one doesn't

**Review by oflatt — CHANGES_REQUESTED** [review]

File: tests/snapshots/files__implicit-return-optimize.snap (hunk 2 above)
Comment (yihozhang) [reply tests/snapshots/files__implicit-return-optimize.snap:56]: seems like we missed the constant folding and computed c9_ as 15 - 1 (v10_ in the new program)

File: dag_in_context/src/utility/util.egg (hunk 1 above)
Comment (oflatt) [reply dag_in_context/src/utility/util.egg:32]: oh great, thx

File: tests/snapshots/files__implicit-return-optimize.snap (hunk 2 above)
Comment (oflatt) [reply tests/snapshots/files__implicit-return-optimize.snap:56]: That does seem like a regression- do you know why it isn't happening now? Perhaps we actually did need `Get` for a generated concat?
Could be that we need to demand `Get` for all of the generated `Concat`, just not subtuples

**Comment by yihozhang** [comment]
I investigated a bit and I have no idea why the optimization is not firing... Some clues:

This is the program in main vs in this branch
image

image

At first I thought it is the loop invariant code motion. Both branches do the LICM over the `pow` function, which eliminates the `Sub` node within `pow` (but not `main`). For `main`, it turns out that if you just do inlining before it hits the first schedule (either with manual inlining or `InlineWithSchedule`, the passthrough optimization alone (maybe together with some other optimizations) will produce the desired program.

It is a bit weird to me why passthrough can do this optimization, because all it does is unioning the output of the loop with the input of the loop, without giving the body any more information. Maybe there's a magic rule in the helpers that passthrough enables that does this... Need to investigate more.

**Comment by oflatt** [comment]
It's it passthrough combined with the interval analysis?

**Comment by yihozhang** [comment]
I also suspected this, but interval analysis does not match on `DoWhile`s, so I can't think of how the combination of the two does it.

**Comment by oflatt** [comment]
Wow, interesting
I think I'm okay with merging this regression, or you can improve interval analysis to be aware of passthrough values in loops

File: tests/snapshots/files__if_dead_code_nested-optimize-sequential.snap [hunk 3]
```diff
@@ -34,10 +34,30 @@ expression: visualization.result
 .b22_:
   v24_: bool = lt c5_ v0;
   c25_: int = const 4;
-  v26_: int = select v24_ c5_ c25_;
```
Comment (yihozhang) [inline tests/snapshots/files__if_dead_code_nested-optimize-sequential.snap:37]: just pulled from main again, this seems like a regression

File: tests/snapshots/files__implicit-return-optimize.snap (hunk 2 above)
Comment (yihozhang) [reply tests/snapshots/files__implicit-return-optimize.snap]: oh nice this is gone after your new IVT is merged

File: tests/snapshots/files__if_dead_code_nested-optimize-sequential.snap (hunk 3 above)
Comment (yihozhang) [reply tests/snapshots/files__if_dead_code_nested-optimize-sequential.snap:37]: nvm, this is sequential

**Review by oflatt — APPROVED** [review]
Looks good!

**Pull request merged**